src/vmem.cc

```diff
@@ -0,0 +1,151 @@
+{
+  num_cpus = number_of_cpus;
+  pt_levels = page_table_levels;
+
+  // calculate number of physical pages
+  page_size = pg_size;
+  if(capacity%page_size != 0)
+    {
+      std::cout << "VirtualMemory initialization error: memory capacity must be a multiple of page size!" << std::endl;
+      exit(0);
+    }
+  num_ppages = capacity/page_size;
+  std::cout << std::endl << "VirtualMemory physical capacity: " << capacity << " num_ppages: " << num_ppages << std::endl;
+
+  // calculate log2_page_size
+  log2_page_size = 0;
+  uint32_t page_size_temp = page_size;
+  for(uint32_t i=0; i<32; i++)
+    {
+      if((page_size_temp&1) == 1)
+	{
+	  log2_page_size = i;
+	  break;
+	}
+      page_size_temp >>= 1;
+    }
+  if((page_size_temp != 1) || (pg_size < 1024))
+    {
+      std::cout<< "VirtualMemory initialization error: page size must be a power of 2, and at least 1024!" << std::endl;
+      exit(0);
+    }
+  std::cout << "VirtualMemory page size: " << page_size << " log2_page_size: " << log2_page_size << std::endl;
+  
+  // initialize random number generator
+  rand_state = random_seed+VMEM_RAND_FACTOR;
+  if(rand_state == 0)
+    {
+      rand_state = VMEM_RAND_FACTOR<<1;
+    }
+
+  std::cout << "VirtualMemory initalizing ppage free list ... ";
+  ppage_free_list.resize(num_ppages);
+  // populate the free list
+  for(uint64_t i=0; i<num_ppages; i++)
+    {
+      ppage_free_list[i] = i;
+    }
+  std::cout << "done" << std::endl << "VirtualMemory shuffling ppage free list ... ";
+  // remove the reserve space from the free list
+  uint64_t num_reserve_ppages = VMEM_RESERVE_CAPACITY/pg_size;
+  if(num_reserve_ppages == 0)
+    {
+      num_reserve_ppages = 1;
+    }
+  for(int i=0; i<num_reserve_ppages; i++)
+    {
+      ppage_free_list.pop_front();
+    }
+  // then shuffle it
+  uint64_t num_swap_ppages = num_ppages-num_reserve_ppages;
+  for(uint64_t i=0; i<num_swap_ppages; i++)
+    {
+      // i is the swap source, swap target is random
+      uint64_t swap_target = (vmem_rand()%(num_swap_ppages));
+
+      uint64_t swap_temp = ppage_free_list[i];
+      ppage_free_list[i] = ppage_free_list[swap_target];
+      ppage_free_list[swap_target] = swap_temp;
+    }
+  std::cout << "done" << std::endl << std::endl;
+
+  // initialize V to P page map tables
+  vpage_to_ppage_map = new std::map<uint64_t, uint64_t>[num_cpus];
+
+  // initialize per-process page tables
+  page_table = new std::map<uint64_t, uint64_t>*[num_cpus];
+  for(uint32_t i=0; i<num_cpus; i++)
+    {
+      page_table[i] = new std::map<uint64_t, uint64_t>[pt_levels];
+    }
+}
+
+VirtualMemory::~VirtualMemory()
+{
+}
+
+uint64_t VirtualMemory::get_next_free_ppage()
+{
+  if(ppage_free_list.empty())
+    {
+      // ran out of physical pages to allocate, so throw error and exit
+      std::cout << "VirtualMemory error: ran out of physical pages to allocate!  Try a larger memory size." << std::endl;
```

Comment:
I would rather leave this for later, if anyone ever requests this functionality.  In order to do it right, we'd have to also do TLB shootdowns and simulate disk access latency.  I think it would be a lot of work that nobody would even appreciate, because so far nobody's using ChampSim for page swapping studies.  I believe the current solution is preferable right now.